app/src/main/java/org/thoughtcrime/securesms/util/SaveAttachmentTask.java

```diff
@@ -0,0 +1,362 @@
+          long total = StreamUtil.copy(inputStream, outputStream);
+          if (total > 0) {
+            updateValues.put(MediaStore.MediaColumns.SIZE, total);
+          }
+        }
+      }
+    }
+
+    if (Build.VERSION.SDK_INT > 28) {
+      updateValues.put(MediaStore.MediaColumns.IS_PENDING, 0);
+    }
+
+    if (updateValues.size() > 0) {
+      getContext().getContentResolver().update(mediaUri, updateValues, null, null);
+    }
+
+    return outputUri.getLastPathSegment();
+  }
+
+  private @NonNull Uri getMediaStoreContentUriForType(@NonNull String contentType) {
+    if (contentType.startsWith("video/")) {
+      return StorageUtil.getVideoUri();
+    } else if (contentType.startsWith("audio/")) {
+      return StorageUtil.getAudioUri();
+    } else if (contentType.startsWith("image/")) {
+      return StorageUtil.getImageUri();
+    } else {
+      return StorageUtil.getDownloadUri();
+    }
+  }
+
+  private @Nullable File ensureExternalPath(@Nullable File path) {
+    if (path != null && path.exists()) {
+      return path;
+    }
+
+    if (path == null) {
+      File documents = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS);
+      if (documents.exists() || documents.mkdirs()) {
+        return documents;
+      } else {
+        return null;
+      }
+    }
+
+    if (path.mkdirs()) {
+      return path;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Returns a path to a shared media (or documents) directory for the type of the file.
+   *
+   * Note that this method attempts to create a directory if the path returned from
+   * Environment object does not exist yet. The attempt may fail in which case it attempts
+   * to return the default "Document" path. It finally returns null if it also fails.
+   * Otherwise it returns the absolute path to the directory.
+   *
+   * @param contentType a MIME type of a file
+   * @return an absolute path to a directory or null
+   */
+  private @Nullable String getExternalPathForType(String contentType) {
+    File storage = null;
+    if (contentType.startsWith("video/")) {
+      storage = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES);
+    } else if (contentType.startsWith("audio/")) {
+      storage = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC);
+    } else if (contentType.startsWith("image/")) {
+      storage = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES);
+    }
+
+    storage = ensureExternalPath(storage);
+    if (storage == null) {
+      return null;
+    }
+
+    return storage.getAbsolutePath();
+  }
+
+  private String generateOutputFileName(@NonNull String contentType, long timestamp) {
+    MimeTypeMap      mimeTypeMap   = MimeTypeMap.getSingleton();
+    String           extension     = mimeTypeMap.getExtensionFromMimeType(contentType);
+    SimpleDateFormat dateFormatter = new SimpleDateFormat("yyyy-MM-dd-HH-mm-ss-SSS");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The SimpleDateFormat object was created without setting the time zone. Make sure that you want to use the default time zone or use letter z, Z or X in the pattern. Otherwise, if you do not call setTimeZone() on the created SimpleDateFormat object, you might get an unexpected date and time when using the object.